Correctly calculate offset to data following Header

wesleywiser opened the pull request:
When an `INumber` value is created, for values other `NumberType::Static` or `NumberType::I24`, a `Header` is allocated with trailing space for the 64-bit data value. However, the calculation used to create the pointer to the data when accessing it via the `as_*_unchecked*` family of functions does not work properly because it first casts the `Header` pointer to a 64-bit value pointer _and then_ does the offset which is incorrect on platforms where the `Header` does not occupy 64-bits such as i686-unknown-linux-gnu.

As a result, the data pointer is incorrectly offset and partially points out of bounds for the allocation. This seems to work ok with some memory allocators (possibly because they round the allocation size up to the next power of 2 and so the adjacent memory is unused) but on others causes heap corruption leading to SIGSEV.

miri detects this situation and offers the following diagnostic:

```shell
$ cargo +nightly miri test --target=i686-unknown-linux-gnu -- can_store_various_numbers
...
running 1 test
test number::tests::can_store_various_numbers ... error: Undefined Behavior: dereferencing pointer failed: alloc135951 has size 12, so pointer to 8 bytes starting at offset 8 is out-of-bounds
   --> src/number.rs:79:9
    |
79  |         &mut *(self as *mut _ as *mut i64).add(1)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ dereferencing pointer failed: alloc135951 has size 12, so pointer to 8 bytes starting at offset 8 is out-of-bounds
    |
    = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
    = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
    = note: BACKTRACE:
    = note: inside `number::Header::as_i64_unchecked_mut` at src/number.rs:79:9
note: inside `number::INumber::new_i64` at src/number.rs:437:18
note: inside `<number::INumber as std::convert::From<i32>>::from` at src/number.rs:598:9
note: inside closure at src/number.rs:711:26
note: inside `number::tests::can_store_various_numbers` at src/number.rs:704:5
```

This corrects the logic used to generate the allocation so that `Header` has a consistent size and alignment.

Diggsey commented:
Thanks for the PR! I'll need some time to make sure I understand the implications of this issue. Would you be able to add a CI step which would detect the problem as well?

Diggsey commented:
Ah, I see the issue. I was expecting that 64-bit values be aligned to 64-bits even on 32-bit architecture. In other words, that the 64-bit offset was always correct.

https://github.com/Diggsey/ijson/blob/0fbe9002ed8edc340ad12fb950adceb793d8f9dc/src/number.rs#L357

Rather than the change in this PR, it might be better to use an approach which ensures the offset matces that in the Layout.

Diggsey commented:
I tested layout generation on `i686-pc-windows-msvc` and it seemed to work correctly, so maybe there is something different about the linux target.

wesleywiser commented:
It looks like preferred alignment on `i686-pc-windows-msvc` is 4 bytes but on `i686-unknown-linux-gnu`, it's 8 bytes:

https://rust.godbolt.org/z/v8x84c1rf

> Would you be able to add a CI step which would detect the problem as well?

It's definitely possible to add miri testing to CI. Unfortunately, there are other miri errors (stacked borrows violations) which aren't addressed by this PR.

Diggsey commented:
> It looks like preferred alignment on i686-pc-windows-msvc is 4 bytes but on i686-unknown-linux-gnu, it's 8 bytes:

That doesn't matter though, what matters is the alignment of a `u64`. Bizarrely that seems to be only 4 bytes on i686-unknown-linux-gnu (see the field offset of `foo` in the printed layout):

https://rust.godbolt.org/z/hGWea4Md8

I would have naively expected the alignment of a u64 to be the same for all i686 targets.

wesleywiser commented:
Ah good call! That happens because the data layout for [`i686-pc-windows-msvc`](https://github.com/rust-lang/rust/blob/83356b78c4ff3e7d84e977aa6143793545967301/compiler/rustc_target/src/spec/i686_pc_windows_msvc.rs#L26-L27) contains `i64:64` which sets the alignment to 64bits but the data layout for [`i686-unknown-linux-gnu`](https://github.com/rust-lang/rust/blob/83356b78c4ff3e7d84e977aa6143793545967301/compiler/rustc_target/src/spec/i686_unknown_linux_gnu.rs#L14-L15) does not contain that and thus the [LLVM default](https://llvm.org/docs/LangRef.html#data-layout) of `i64:32:64` is used instead. Clang uses the same data layout strings that rustc does so this seems to be expected platform behavior.

Diggsey commented:
So there's two solutions here:

1) Simply always use 8-byte alignment for the payload. We just need to update the layout calculation on the line I linked to before.

2) Make the alignment match the alignment of a u64 on the target. We need to change the offset code to apply the correct offset in the correct circumstances, without adding a performance cost. This is a bit tricky because many of the relevant methods are not "const" yet.

Diggsey commented:
> It's definitely possible to add miri testing to CI. Unfortunately, there are other miri errors (stacked borrows violations) which aren't addressed by this PR.

I already *do* MIRI testing in CI and it passes? I just don't test on 32-bit atm.

wesleywiser commented:
Weird, even using 64-bit Linux, I get the following stacked borrows error (on a clean checkout of master):

```
$ cargo +nightly miri test --target=x86_64-unknown-linux-gnu -- can_store_various_numbers
running 1 test
test number::tests::can_store_various_numbers ... error: Undefined Behavior: trying to retag from <324295> for Unique permission at alloc135765[0x8], but that tag does not exist in the borrow stack for this location
   --> src/number.rs:80:9
    |
80  |         &mut *(self as *mut _ as *mut i64).add(1)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |         |
    |         trying to retag from <324295> for Unique permission at alloc135765[0x8], but that tag does not exist in the borrow stack for this location
    |         this error occurs as part of retag at alloc135765[0x8..0x10]
    |
    = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
    = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information

$ rustc +nightly --version -v
rustc 1.67.0-nightly (77e57db38 2022-10-30)
```

Diggsey commented:
It's possible that MIRI has changed since the last time it was run on this repo - I just approved CI on this PR so we'll soon see.

wesleywiser commented:
Do you have a preference between solutions 1&2? The other option I see would be to explicitly add a `u32` padding field to `Header` and that by itself also fixes the issue.

Diggsey commented:
> Do you have a preference between solutions 1&2?

I think 1 is simpler even if it does use a bit of extra memory on 32-bit linux, so I'd probably go for that.

> The other option I see would be to explicitly add a u32 padding field to Header and that by itself also fixes the issue.

You have to remember that `Header` is used for `Static` and `Fixed` variants where we don't want the extra padding (because there *is* no payload), so adding another field is not workable.

Diggsey commented:
It looks like MIRI has become more strict, and that's why it's failing. I'll see if I can figure out why.

Diggsey commented:
So, at least in the first example, it's failing because I'm sharing the pointer calculation code between the *const and *mut cases, and it doesn't like that, so it's easily fixable by duplicating the code. It's a shame there's no way to abstract it though...

wesleywiser commented:
Implemented option 1 and confirmed that miri seems ok with it and my `i686-unknown-linux-musl` build no longer SIGSEVs when running tests either. I made the change for all of `I64`, `U64` and `F64` as according to the target data layout, they all have the same alignments. 

I also fixed the clippy warnings.

Diggsey commented:
LGTM, I'll fix the MIRI issues in a separate change.

wesleywiser commented:
Fixed the formatting and satiated clippy by implementing `Eq`. Thanks for the quick turn around on this! 🥇

Diggsey commented:
Thanks!

Diggsey commented:
I fixed the Miri issues and added CI for 32-bit linux and windows. It turned out to be a bit more involved than I thought 😅
